### PR TITLE
Build TCP connection, packet framing, RLPx, DevP2P Wire Protocol and Eth Wire Protocol

### PR DESCRIPTION
**Pull request by hayesgm**
This patch starts to build TCP sessions and packet framing into ExWire. Once we get done framing (and protocol negotitation), we'll have a more or less working RLPx implementation (though, there's lot of details to fill in). Currently, a lot of code is dumped where it needs to be to get things to work (e.g. TCP should really call our to stub code). Additionally, we're only just beginning to get the framing code to work (the spec is, yet again, under-documented or strictly incorrect, so we're basing it off a mix of Geth and Parity's code). Luckily, geth included a test case so that's good. All in all, at this point, we can connect to a TCP node, negotiate keys and try to send packets. The key exchange (ECIES) looks to be right based on test cases, but I'm not 100% positive the derived keys are correct yet (see the TCP module). Luckily, the clients are pretty happy to accept and report any isuses, so we'll be able to diagnose anything that's incorrect after we finish our framing and de-framing code.

### CODE REVIEW

**Comment by hayesgm** [comment]
@masonforest This should be ready to review. We can now connect and pull blocks from a DevP2P peer :D :D :D :D

File: lib/ex_wire/framing/frame.ex [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  """
+
+  alias ExthCrypto.MAC
+  alias ExthCrypto.AES
+  alias ExWire.Framing.Secrets
+
+  @type frame :: binary()
+
+  @spec frame(integer(), ExRLP.t, Secrets.t) :: {frame, Secrets.t}
+  def frame(packet_type, packet_data, frame_secrets=%Secrets{egress_mac: egress_mac, encoder_stream: encoder_stream, mac_encoder: mac_encoder, mac_secret: mac_secret}) do
```
Comment (masonforest) [inline lib/ex_wire/framing/frame.ex:19]: Is there any way to break this up a little more?

Maybe pull some of the the padding stuff out into a helper?

File: lib/ex_wire/adapter/tcp.ex [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    :ok
+  end
+
+  @doc """
+  Client function to send HELLO message after connecting.
+  """
+  def send_hello(pid) do
+    send_packet(pid, %Packet.Hello{
+      p2p_version: 0x04,
+      client_id: "Exthereum/0.1",
```
Comment (masonforest) [inline lib/ex_wire/adapter/tcp.ex]: Pull this and line 253 into environment variables?

**Review by masonforest — APPROVED** [review]
Looks good!

Looks like `lib/ex_wire/.DS_Store` slipped by. You could probably add that to your global gitignore to avoid those getting in at all.


I see now why this is frustrating. It seems like there's lot of unavoidable complexity in here. I'm not sure it's necessary but I also don't see how we can avoid implementing it.

Some of the longer methods could probably be broken up some more but feel free to merge to get this working an we can refactor later.

Nice work! 💯

File: lib/ex_wire/adapter/tcp.ex (hunk 2 above)
Comment (hayesgm) [reply lib/ex_wire/adapter/tcp.ex]: Yeah, definitely going to do some minor re-factors on this stuff.

File: lib/ex_wire/framing/frame.ex (hunk 1 above)
Comment (hayesgm) [reply lib/ex_wire/framing/frame.ex:19]: I can probably do so-- this code was supposed to be based off the spec comments, but it ended up getting built off of translating code from Geth / Parity, hence why it doesn't feel very idiomatic. I can probably fix that a bunch.

**Comment by hayesgm** [comment]
Yeah, I'll merge now and then go back and clean up some of the functions. It'll be nice to have this mostly done and get to move on to more fun things (like syncing the entire blockchain :))

**Pull request merged**